lib/time.c

```diff
@@ -241,19 +287,38 @@ grn_str2timeval(const char *str, uint32_t str_len, grn_timeval *tv)
     return GRN_INVALID_ARGUMENT;
   }
   r1 = r2;
+
+  if ((r1 + 1) < rend && *r1 == '.') {
+    uv = grn_atoi(++r1, rend, &r2);
+
+    for (int i = 0; r2 + i < r1 + 6; i++) {
+      uv *= 10;
+    }
+    if (uv >= GRN_TIME_USEC_PER_SEC) { return GRN_INVALID_ARGUMENT; }
+  }
+
   tm.tm_yday = -1;
   tm.tm_isdst = -1;
 
-  /* tm_yday is set appropriately (0-365) on successful completion. */
-  tv->tv_sec = grn_mktime(&tm);
-  if (tm.tm_yday == -1) { return GRN_INVALID_ARGUMENT; }
-  if ((r1 + 1) < rend && *r1 == '.') { r1++; }
-  uv = grn_atoi(r1, rend, &r2);
-  while (r2 < r1 + 6) {
-    uv *= 10;
-    r2++;
+  {
+    bool have_offset = false;
+    int32_t gm_offset_sec = 0;
+    grn_rc rc = GRN_SUCCESS;
+
+    rc = grn_str2timeval_offset_sec(r2, rend, &have_offset, &gm_offset_sec);
+
+    if (rc != GRN_SUCCESS) { return rc; }
+
+    if (have_offset) {
+      tv->tv_sec = grn_timegm(&tm);
+      tv->tv_sec -= gm_offset_sec;
+    } else {
+      tv->tv_sec = grn_mktime(&tm);
+      /* tm_yday is set appropriately (0-365) on successful completion. */
+      if (tm.tm_yday == -1) { return GRN_INVALID_ARGUMENT; }
+    }
```

Comment:
I think that we also need this for `grn_timegm()` case.

```suggestion
    }
    /* tm_yday is set appropriately (0-365) on successful completion. */
    if (tm.tm_yday == -1) { return GRN_INVALID_ARGUMENT; }
```

Reply:
I have fixed it.